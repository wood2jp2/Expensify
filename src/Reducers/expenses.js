// Expenses Reducer and default values
const expensesReducerDefaultState = []

// The action is the result of the action functions
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
            return [...state, action.expense]
        case "REMOVE_EXPENSE":
            return state.filter(({ id }) => id !== action.id)
        case "EDIT_EXPENSE":
            return state.map(expense => {
                return expense.id === action.id ?
                    {...expense, ...action.updates} :
                    expense
            })
        default: 
            return state 
    }
}