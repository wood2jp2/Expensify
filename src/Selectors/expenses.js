import moment from 'moment'

export default (expensesState, {text, sortBy, startDate, endDate}) => {
    return expensesState.filter(expense => {
        const 
             createdAtMoment = moment(expense.createdAt),
             startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true,
             endDateMatch  = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true,
             textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        // Should be sorted most recent expense to oldest expense
        if (sortBy === 'date') {
            return b.createdAt - a.createdAt 
        } 
        // Should be sorted most expensive to least expensive
        else if (sortBy === 'amount') {
            return b.amount - a.amount
        }
    })
}