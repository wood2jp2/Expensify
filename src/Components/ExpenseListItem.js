import React from 'react'
import { connect } from 'react-redux'
import { removeExpense } from '../Actions/expenses'

const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
    <div>
        <h3>{description}</h3>
        <p>Amount: {amount}</p>
        <p>Created At: {createdAt}</p>
        <button onClick={ () => {
            dispatch(removeExpense({id}))
        }}>Remove</button>
    </div>
)

const ConnectedExpenseListItem = connect()(ExpenseListItem)

export default ConnectedExpenseListItem