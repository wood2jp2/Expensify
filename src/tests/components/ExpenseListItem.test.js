import React from 'react'
import { shallow } from 'enzyme'
import { expenses } from '../fixtures/expenses'
import ExpenseListItem from '../../Components/ExpenseListItem'

test("should render an expenseListItem", () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>)

    expect(wrapper).toMatchSnapshot()
    // expect(wrapper.instance().props).toBe(195)
})