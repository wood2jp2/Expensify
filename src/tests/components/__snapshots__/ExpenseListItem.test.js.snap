// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expenseListItem 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/edit/1"
    >
      Gum
    </Link>
  </h3>
  <p>
    Amount: 
    195
  </p>
  <p>
    Created At: 
    0
  </p>
</div>
`;
